File: js/src/image.ts
import {
  GenericResult,
  GenericResult_GenericStatus,
  RegistryAuthType,
  ImageRegistryConfig,
  Image as ImageProto,
  GPUConfig,
} from "../proto/modal_proto/api";
import { type ModalClient } from "./client";
import { App, parseGpuConfig } from "./app";
import { Secret, mergeEnvIntoSecrets } from "./secret";
import { ClientError } from "nice-grpc";
import { Status } from "nice-grpc";
import { NotFoundError, InvalidError } from "./errors";
import { checkObjectName } from "./name_utils";

const DEFAULT_IMAGE_TAG = "latest";

function validateImageName(name: string): void {
  checkObjectName(name, "Image");
  if (name.startsWith("im-")) {
    throw new InvalidError(
      "Image name cannot start with 'im-' (reserved for image IDs).",
    );
  }
}

function validateImageTag(tag: string): void {
  checkObjectName(tag, "Image tag");
}

/**
 * Parse an image reference, returning [namespacePrefix, nameTag].
 *
 * If the name contains a '/', the part before the last '/' is extracted as
 * a namespace prefix (intended for environment/name or workspace/env/name
 * syntax). The actual image name (after the last '/') is validated as a
 * standard image name.
 */
function parseNamedImageRef(value: string): [string, string] {
  const separatorIndex = value.indexOf(":");
  let imageName: string;
  let tag: string;
  if (separatorIndex === -1) {
    imageName = value;
    tag = DEFAULT_IMAGE_TAG;
  } else {
    imageName = value.slice(0, separatorIndex);
    tag = value.slice(separatorIndex + 1);
  }

  let prefix = "";
  const lastSlashIndex = imageName.lastIndexOf("/");
  if (lastSlashIndex !== -1) {
    prefix = imageName.slice(0, lastSlashIndex);
    const after = imageName.slice(lastSlashIndex + 1);
    if (prefix === "") {
      throw new InvalidError(
        "Invalid Image name: '/' prefix must be non-empty.",
      );
    }
    if (after === "") {
      throw new InvalidError(
        "Invalid Image name: name after '/' must be non-empty.",
      );
    }
    imageName = after;
  }

  validateImageName(imageName);
  validateImageTag(tag);

  const fullName = prefix ? `${prefix}/${imageName}` : imageName;
  return [prefix, `${fullName}:${tag}`];
}

/** Optional parameters for {@link ImageService#fromName client.images.fromName()}. */
export type ImageFromNameParams = {
  /** Modal Environment to resolve the named Image in. */
  environment?: string;
};

/**
 * Service for managing {@link Image}s.
 *
 * Normally only ever accessed via the client as:
 * ```typescript
 * const modal = new ModalClient();
 * const image = await modal.images.fromRegistry("alpine");
 * ```
 */
export class ImageService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Creates an {@link Image} from an Image ID
   *
   * @param imageId - Image ID.
   */
  async fromId(imageId: string): Promise<Image> {
    try {
      const resp = await this.#client.cpClient.imageFromId({ imageId });
      return new Image(this.#client, resp.imageId, "");
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(err.details);
      if (
        err instanceof ClientError &&
        err.code === Status.FAILED_PRECONDITION &&
        err.details.includes("Could not find image with ID")
      )
        throw new NotFoundError(err.details);
      throw err;
    }
  }

  /**
   * Reference a named {@link Image} that was previously published.
   *
   * @param name - Name of the published Image, optionally including a tag as `name:tag`.
   * If no tag is included, `:latest` is used.
   * @param params - Optional environment.
   */
  async fromName(
    name: string,
    params: ImageFromNameParams = {},
  ): Promise<Image> {
    const [namespacePrefix, tag] = parseNamedImageRef(name);

    let environmentName: string;
    if (namespacePrefix) {
      if (params.environment) {
        throw new InvalidError(
          "Cannot specify 'environment' when the image name contains a '/'.",
        );
      }
      environmentName = "";
    } else {
      environmentName = this.#client.environmentName(params.environment);
    }

    try {
      const resp = await this.#client.cpClient.imageGetByTag({
        environmentName,
        tag,
      });
      return new Image(this.#client, resp.imageId, "");
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(err.details);
      throw err;
    }
  }

  /**
   * Creates an {@link Image} from a raw registry tag, optionally using a {@link Secret} for authentication.
   *
   * @param tag - The registry tag for the Image.
   * @param secret - Optional. A Secret containing credentials for registry authentication.
   */
  fromRegistry(tag: string, secret?: Secret): Image {
    let imageRegistryConfig;
    if (secret) {
      if (!(secret instanceof Secret)) {
        throw new TypeError(
          "secret must be a reference to an existing Secret, e.g. `await client.secrets.fromName('my_secret')`",
        );
      }
      imageRegistryConfig = {
        registryAuthType: RegistryAuthType.REGISTRY_AUTH_TYPE_STATIC_CREDS,
        secretId: secret.secretId,
      };
    }
    return new Image(this.#client, "", tag, imageRegistryConfig);
  }

  /**
   * Creates an {@link Image} from a raw registry tag, optionally using a {@link Secret} for authentication.
   *
   * @param tag - The registry tag for the Image.
   * @param secret - A Secret containing credentials for registry authentication.
   */
  fromAwsEcr(tag: string, secret: Secret): Image {
    let imageRegistryConfig;
    if (secret) {
      if (!(secret instanceof Secret)) {
        throw new TypeError(
          "secret must be a reference to an existing Secret, e.g. `await client.secrets.fromName('my_secret')`",
        );
      }
      imageRegistryConfig = {
        registryAuthType: RegistryAuthType.REGISTRY_AUTH_TYPE_AWS,
        secretId: secret.secretId,
      };
    }
    return new Image(this.#client, "", tag, imageRegistryConfig);
  }

  /**
   * Creates an {@link Image} from a raw registry tag, optionally using a {@link Secret} for authentication.
   *
   * @param tag - The registry tag for the Image.
   * @param secret - A Secret containing credentials for registry authentication.
   */
  fromGcpArtifactRegistry(tag: string, secret: Secret): Image {
    let imageRegistryConfig;
    if (secret) {
      if (!(secret instanceof Secret)) {
        throw new TypeError(
          "secret must be a reference to an existing Secret, e.g. `await client.secrets.fromName('my_secret')`",
        );
      }
      imageRegistryConfig = {
        registryAuthType: RegistryAuthType.REGISTRY_AUTH_TYPE_GCP,
        secretId: secret.secretId,
      };
    }
    return new Image(this.#client, "", tag, imageRegistryConfig);
  }

  /**
   * Delete an {@link Image} by ID.
   *
   * Deletion is irreversible and will prevent Functions/Sandboxes from using the Image.
   *
   * Note: When building an Image, each chained method call will create an
   * intermediate Image layer, each with its own ID. Deleting an Image will not
   * delete any of its intermediate layers, only the image identified by the
   * provided ID.
   */
  async delete(imageId: string, _: ImageDeleteParams = {}): Promise<void> {
    try {
      await this.#client.cpClient.imageDelete({ imageId });
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(err.details);
      if (
        err instanceof ClientError &&
        err.code === Status.FAILED_PRECONDITION &&
        err.details.includes("Could not find image with ID")
      )
        throw new NotFoundError(err.details);
      throw err;
    }
  }
}

/** Optional parameters for {@link ImageService#delete client.images.delete()}. */
export type ImageDeleteParams = Record<never, never>;

/** Optional parameters for {@link Image#publish Image.publish()}. */
export type ImagePublishParams = {
  /** Modal Environment to publish the named Image in. */
  environment?: string;
};

/** Optional parameters for {@link Image#dockerfileCommands Image.dockerfileCommands()}. */
export type ImageDockerfileCommandsParams = {
  /** Environment variables to set in the build environment. */
  env?: Record<string, string>;

  /** {@link Secret}s that will be made available as environment variables to this layer's build environment. */
  secrets?: Secret[];

  /** GPU reservation for this layer's build environment (e.g. "A100", "T4:2", "A100-80GB:4"). */
  gpu?: string;

  /** Ignore cached builds for this layer, similar to 'docker build --no-cache'. */
  forceBuild?: boolean;
};

/** Represents a single image layer with its build configuration. */
type Layer = {
  commands: string[];
  env?: Record<string, string>;
  secrets?: Secret[];
  gpuConfig?: GPUConfig;
  forceBuild?: boolean;
};

/** A container image, used for starting {@link Sandbox}es. */
export class Image {
  #client: ModalClient;
  #imageId: string;
  #tag: string;
  // Image ID of the parent image layer to use as `FROM base`.
  #baseImageId: string;
  #imageRegistryConfig?: ImageRegistryConfig;
  #layers: Layer[];

  /** @ignore */
  constructor(
    client: ModalClient,
    imageId: string,
    tag: string,
    imageRegistryConfig?: ImageRegistryConfig,
    layers?: Layer[],
    baseImageId?: string,
  ) {
    this.#client = client;
    this.#imageId = imageId;
    this.#tag = tag;
    this.#baseImageId = baseImageId || "";
    this.#imageRegistryConfig = imageRegistryConfig;
    this.#layers = layers || [
      {
        commands: [],
        env: undefined,
        secrets: undefined,
        gpuConfig: undefined,
        forceBuild: false,
      },
    ];
  }
  get imageId(): string {
    return this.#imageId;
  }

  private static validateDockerfileCommands(commands: string[]): void {
    for (const command of commands) {
      const trimmed = command.trim().toUpperCase();
      if (trimmed.startsWith("COPY ") && !trimmed.startsWith("COPY --FROM=")) {
        throw new InvalidError(
          "COPY commands that copy from local context are not yet supported.",
        );
      }
    }
  }

  /**
   * Extend an image with arbitrary Dockerfile-like commands.
   *
   * Each call creates a new Image layer that will be built sequentially.
   * The provided options apply only to this layer.
   *
   * @param commands - Array of Dockerfile commands as strings
   * @param params - Optional configuration for this layer's build
   * @returns A new Image instance
   */
  dockerfileCommands(
    commands: string[],
    params?: ImageDockerfileCommandsParams,
  ): Image {
    if (commands.length === 0) {
      return this;
    }

    Image.validateDockerfileCommands(commands);

    const newLayer: Layer = {
      commands: [...commands],
      env: params?.env,
      secrets: params?.secrets,
      gpuConfig: params?.gpu ? parseGpuConfig(params.gpu) : undefined,
      forceBuild: params?.forceBuild,
    };

    const baseImageId = this.#imageId || this.#baseImageId;
    const layers = this.#imageId === "" ? this.#layers : [];

    return new Image(
      this.#client,
      "",
      this.#tag,
      this.#imageRegistryConfig,
      [...layers, newLayer],
      baseImageId,
    );
  }

  /**
   * Eagerly builds an Image on Modal.
   *
   * @param app - App to use to build the Image.
   */
  async build(app: App): Promise<Image> {
    if (this.imageId !== "") {
      // Image is already built with an Image ID
      return this;
    }

    this.#client.logger.debug("Building image", "app_id", app.appId);

    let baseImageId: string | undefined = this.#baseImageId || undefined;
    const imageBuilderVersion = await this.#client.getImageBuilderVersion(
      app.environmentName,
    );

    for (let i = 0; i < this.#layers.length; i++) {
      const layer = this.#layers[i];

      const mergedSecrets = await mergeEnvIntoSecrets(
        this.#client,
        layer.env,
        layer.secrets,
      );

      const secretIds = mergedSecrets.map((secret) => secret.secretId);
      const gpuConfig = layer.gpuConfig;

      let dockerfileCommands: string[];
      let baseImages: Array<{ dockerTag: string; imageId: string }>;

      if (i === 0 && baseImageId) {
        dockerfileCommands = ["FROM base", ...layer.commands];
        baseImages = [{ dockerTag: "base", imageId: baseImageId }];
      } else if (i === 0) {
        dockerfileCommands = [`FROM ${this.#tag}`, ...layer.commands];
        baseImages = [];
      } else {
        dockerfileCommands = ["FROM base", ...layer.commands];
        baseImages = [{ dockerTag: "base", imageId: baseImageId! }];
      }

      const imageRegistryConfig =
        i === 0 && !baseImageId ? this.#imageRegistryConfig : undefined;

      const resp = await this.#client.cpClient.imageGetOrCreate({
        appId: app.appId,
        image: ImageProto.create({
          dockerfileCommands,
          imageRegistryConfig,
          secretIds,
          gpuConfig,
          contextFiles: [],
          baseImages,
        }),
        builderVersion: imageBuilderVersion,
        forceBuild: layer.forceBuild || false,
      });

      let result: GenericResult;

      if (resp.result?.status) {
        // Image has already been built
        result = resp.result;
      } else {
        // Not built or in the process of building - wait for build
        let lastEntryId = "";
        let resultJoined: GenericResult | undefined = undefined;
        while (!resultJoined) {
          for await (const item of this.#client.cpClient.imageJoinStreaming({
            imageId: resp.imageId,
            timeout: 55,
            lastEntryId,
          })) {
            if (item.entryId) lastEntryId = item.entryId;
            if (item.result?.status) {
              resultJoined = item.result;
              break;
            }
            // Ignore all log lines and progress updates.
          }
        }
        result = resultJoined;
      }

      if (
        result.status === GenericResult_GenericStatus.GENERIC_STATUS_FAILURE
      ) {
        throw new Error(
          `Image build for ${resp.imageId} failed with the exception:\n${result.exception}`,
        );
      } else if (
        result.status === GenericResult_GenericStatus.GENERIC_STATUS_TERMINATED
      ) {
        throw new Error(
          `Image build for ${resp.imageId} terminated due to external shut-down. Please try again.`,
        );
      } else if (
        result.status === GenericResult_GenericStatus.GENERIC_STATUS_TIMEOUT
      ) {
        throw new Error(
          `Image build for ${resp.imageId} timed out. Please try again with a larger timeout parameter.`,
        );
      } else if (
        result.status !== GenericResult_GenericStatus.GENERIC_STATUS_SUCCESS
      ) {
        throw new Error(
          `Image build for ${resp.imageId} failed with unknown status: ${result.status}`,
        );
      }

      // the new image is the base for the next layer
      baseImageId = resp.imageId;
    }
    this.#imageId = baseImageId!;
    this.#client.logger.debug("Image build completed", "image_id", baseImageId);
    return this;
  }

  /**
   * Publish this built Image under a stable name and tag.
   *
   * @param name - Name to publish the Image under, optionally including a tag as `name:tag`.
   * If no tag is included, `:latest` is used.
   * @param params - Optional environment.
   */
  async publish(name: string, params: ImagePublishParams = {}): Promise<void> {
    const [namespacePrefix, tag] = parseNamedImageRef(name);

    if (this.#imageId === "") {
      throw new InvalidError(
        "Cannot publish an image that has not been built yet. Call build() first.",
      );
    }

    let environmentName: string;
    if (namespacePrefix) {
      if (params.environment) {
        throw new InvalidError(
          "Cannot specify 'environment' when the image name contains a '/'.",
        );
      }
      environmentName = "";
    } else {
      environmentName = this.#client.environmentName(params.environment);
    }

    await this.#client.cpClient.imagePublish({
      imageId: this.#imageId,
      environmentName,
      allowPublic: false,
      tag,
    });
  }
}

File: go/image.go
package modal

import (
	"context"
	"fmt"
	"io"
	"strings"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// ImageService provides Image related operations.
type ImageService interface {
	FromRegistry(tag string, params *ImageFromRegistryParams) *Image
	FromAwsEcr(tag string, secret *Secret, params *ImageFromAwsEcrParams) *Image
	FromGcpArtifactRegistry(tag string, secret *Secret, params *ImageFromGcpArtifactRegistryParams) *Image
	FromID(ctx context.Context, imageID string, params *ImageFromIDParams) (*Image, error)
	FromName(ctx context.Context, name string, params *ImageFromNameParams) (*Image, error)
	Delete(ctx context.Context, imageID string, params *ImageDeleteParams) error
}

type imageServiceImpl struct{ client *Client }

const defaultImageTag = "latest"

func validateImageName(name string) error {
	if err := checkObjectName(name, "Image"); err != nil {
		return err
	}
	if strings.HasPrefix(name, "im-") {
		return InvalidError{Exception: "Image name cannot start with 'im-' (reserved for image IDs)."}
	}
	return nil
}

func validateImageTag(tag string) error {
	return checkObjectName(tag, "Image tag")
}

// parseNamedImageRef parses an image reference, returning (namespacePrefix, nameTag).
// If the name contains a '/', the part before the last '/' is extracted as a
// namespace prefix (intended for environment/name or workspace/env/name syntax).
// The actual image name (after the last '/') is validated as a standard image name.
func parseNamedImageRef(value string) (string, string, error) {
	imageName, tag, found := strings.Cut(value, ":")
	if !found {
		tag = defaultImageTag
	}

	prefix := ""
	if lastSlash := strings.LastIndex(imageName, "/"); lastSlash != -1 {
		prefix = imageName[:lastSlash]
		after := imageName[lastSlash+1:]
		if prefix == "" {
			return "", "", InvalidError{Exception: "Invalid Image name: '/' prefix must be non-empty."}
		}
		if after == "" {
			return "", "", InvalidError{Exception: "Invalid Image name: name after '/' must be non-empty."}
		}
		imageName = after
	}

	if err := validateImageName(imageName); err != nil {
		return "", "", err
	}
	if err := validateImageTag(tag); err != nil {
		return "", "", err
	}

	fullName := imageName
	if prefix != "" {
		fullName = prefix + "/" + imageName
	}
	return prefix, fullName + ":" + tag, nil
}

// ImageDockerfileCommandsParams are options for Image.DockerfileCommands().
type ImageDockerfileCommandsParams struct {
	// Environment variables to set in the build environment.
	Env map[string]string

	// Secrets that will be made available as environment variables to this layer's build environment.
	Secrets []*Secret

	// GPU reservation for this layer's build environment (e.g. "A100", "T4:2", "A100-80GB:4").
	GPU string

	// Ignore cached builds for this layer, similar to 'docker build --no-cache'.
	ForceBuild bool
}

// layer represents a single image layer with its build configuration.
type layer struct {
	commands   []string
	env        map[string]string
	secrets    []*Secret
	gpu        string
	forceBuild bool
}

// Image represents a Modal Image, which can be used to create Sandboxes.
type Image struct {
	ImageID string

	// buildRegistryConfig, when non-nil, hydrates the registry-auth Secret (if
	// any) and returns the ImageRegistryConfig for the base build layer. It runs
	// at Build time rather than when the Image is constructed, since the Secret
	// may be lazy (created via FromMap). nil when not pulling from a private registry.
	buildRegistryConfig func(ctx context.Context, client *Client) (*pb.ImageRegistryConfig, error)
	tag                 string
	// baseImageID is the image ID of the parent image layer to use as FROM base.
	baseImageID string
	layers      []layer

	client *Client
}

// ImageFromRegistryParams are options for creating an Image from a registry.
type ImageFromRegistryParams struct {
	Secret *Secret
}

// ImageFromAwsEcrParams are options for ImageService.FromAwsEcr.
type ImageFromAwsEcrParams struct{}

// ImageFromGcpArtifactRegistryParams are options for ImageService.FromGcpArtifactRegistry.
type ImageFromGcpArtifactRegistryParams struct{}

// ImageFromIDParams are options for ImageService.FromID.
type ImageFromIDParams struct{}

// ImageFromNameParams are options for ImageService.FromName.
type ImageFromNameParams struct {
	Environment string
}

// ImageBuildParams are options for Image.Build.
type ImageBuildParams struct{}

// ImagePublishParams are options for Image.Publish.
type ImagePublishParams struct {
	Environment string
}

// registryConfigBuilder returns a closure that hydrates the registry-auth Secret
// (if any) and builds the ImageRegistryConfig. It runs at Build time so that lazy
// Secrets (created via FromMap) are resolved against the build's client.
func registryConfigBuilder(authType pb.RegistryAuthType, secret *Secret) func(ctx context.Context, client *Client) (*pb.ImageRegistryConfig, error) {
	return func(ctx context.Context, client *Client) (*pb.ImageRegistryConfig, error) {
		secretID := ""
		if secret != nil {
			if err := hydrateSecrets(ctx, client, []*Secret{secret}); err != nil {
				return nil, err
			}
			secretID = secret.SecretID
		}
		return pb.ImageRegistryConfig_builder{
			RegistryAuthType: authType,
			SecretId:         secretID,
		}.Build(), nil
	}
}

// FromRegistry builds a Modal Image from a public or private image registry without any changes.
func (s *imageServiceImpl) FromRegistry(tag string, params *ImageFromRegistryParams) *Image {
	image := &Image{
		ImageID: "",
		tag:     tag,
		layers:  []layer{{}},
		client:  s.client,
	}
	if params != nil && params.Secret != nil {
		image.buildRegistryConfig = registryConfigBuilder(pb.RegistryAuthType_REGISTRY_AUTH_TYPE_STATIC_CREDS, params.Secret)
	}
	return image
}

// FromAwsEcr creates an Image from an AWS ECR tag
func (s *imageServiceImpl) FromAwsEcr(tag string, secret *Secret, params *ImageFromAwsEcrParams) *Image {
	return &Image{
		ImageID:             "",
		buildRegistryConfig: registryConfigBuilder(pb.RegistryAuthType_REGISTRY_AUTH_TYPE_AWS, secret),
		tag:                 tag,
		layers:              []layer{{}},
		client:              s.client,
	}
}

// FromGcpArtifactRegistry creates an Image from a GCP Artifact Registry tag.
func (s *imageServiceImpl) FromGcpArtifactRegistry(tag string, secret *Secret, params *ImageFromGcpArtifactRegistryParams) *Image {
	return &Image{
		ImageID:             "",
		buildRegistryConfig: registryConfigBuilder(pb.RegistryAuthType_REGISTRY_AUTH_TYPE_GCP, secret),
		tag:                 tag,
		layers:              []layer{{}},
		client:              s.client,
	}
}

// FromID looks up an Image from an ID
func (s *imageServiceImpl) FromID(ctx context.Context, imageID string, params *ImageFromIDParams) (*Image, error) {
	resp, err := s.client.cpClient.ImageFromId(
		ctx,
		pb.ImageFromIdRequest_builder{
			ImageId: imageID,
		}.Build(),
	)
	if status, ok := status.FromError(err); ok && status.Code() == codes.NotFound {
		return nil, NotFoundError{fmt.Sprintf("Image '%s' not found", imageID)}
	}
	if err != nil {
		return nil, err
	}

	return &Image{
		ImageID: resp.GetImageId(),
		layers:  []layer{{}},
		client:  s.client,
	}, nil
}

// FromName references a named Image that was previously published.
// The name may include a tag as name:tag; if no tag is included, :latest is used.
func (s *imageServiceImpl) FromName(ctx context.Context, name string, params *ImageFromNameParams) (*Image, error) {
	if params == nil {
		params = &ImageFromNameParams{}
	}
	namespacePrefix, tag, err := parseNamedImageRef(name)
	if err != nil {
		return nil, err
	}

	var environmentName string
	if namespacePrefix != "" {
		if params.Environment != "" {
			return nil, InvalidError{Exception: "Cannot specify 'Environment' when the image name contains a '/'."}
		}
		environmentName = ""
	} else {
		environmentName = firstNonEmpty(params.Environment, s.client.profile.Environment)
	}

	resp, err := s.client.cpClient.ImageGetByTag(ctx, pb.ImageGetByTagRequest_builder{
		EnvironmentName: environmentName,
		Tag:             tag,
	}.Build())
	if st, ok := status.FromError(err); ok && st.Code() == codes.NotFound {
		return nil, NotFoundError{fmt.Sprintf("Image '%s' not found", tag)}
	}
	if err != nil {
		return nil, err
	}

	return &Image{
		ImageID: resp.GetImageId(),
		layers:  []layer{{}},
		client:  s.client,
	}, nil
}

// DockerfileCommands extends an image with arbitrary Dockerfile-like commands.
//
// Each call creates a new Image layer that will be built sequentially.
// The provided options apply only to this layer.
func (image *Image) DockerfileCommands(commands []string, params *ImageDockerfileCommandsParams) *Image {
	if len(commands) == 0 {
		return image
	}

	if params == nil {
		params = &ImageDockerfileCommandsParams{}
	}

	newLayer := layer{
		commands:   append([]string{}, commands...),
		env:        params.Env,
		secrets:    params.Secrets,
		gpu:        params.GPU,
		forceBuild: params.ForceBuild,
	}

	baseImageID := image.baseImageID
	newLayers := append([]layer{}, image.layers...)
	if image.ImageID != "" {
		baseImageID = image.ImageID
		newLayers = []layer{}
	}
	newLayers = append(newLayers, newLayer)

	return &Image{
		ImageID:             "",
		tag:                 image.tag,
		baseImageID:         baseImageID,
		buildRegistryConfig: image.buildRegistryConfig,
		layers:              newLayers,
		client:              image.client,
	}
}

func validateDockerfileCommands(commands []string) error {
	for _, command := range commands {
		trimmed := strings.ToUpper(strings.TrimSpace(command))
		if strings.HasPrefix(trimmed, "COPY ") && !strings.HasPrefix(trimmed, "COPY --FROM=") {
			return InvalidError{"COPY commands that copy from local context are not yet supported."}
		}
	}
	return nil
}

// waitForBuildIteration performs a single iteration of waiting for an image build to complete.
// It streams build updates and returns either a result (if build completes) or nil (if the stream
// ends without a result, requiring another iteration).
func (image *Image) waitForBuildIteration(ctx context.Context, imageID string, lastEntryID string) (*pb.GenericResult, string, error) {
	streamCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	stream, err := image.client.cpClient.ImageJoinStreaming(streamCtx, pb.ImageJoinStreamingRequest_builder{
		ImageId:     imageID,
		Timeout:     55,
		LastEntryId: lastEntryID,
	}.Build())
	if err != nil {
		return nil, lastEntryID, err
	}

	for {
		item, err := stream.Recv()
		if err == io.EOF {
			return nil, lastEntryID, nil
		} else if err != nil {
			return nil, lastEntryID, err
		}

		if item.GetEntryId() != "" {
			lastEntryID = item.GetEntryId()
		}

		// Ignore all log lines and progress updates.
		res := item.GetResult()
		if res == nil || res.GetStatus() == pb.GenericResult_GENERIC_STATUS_UNSPECIFIED {
			continue
		}

		return res, lastEntryID, nil
	}
}

// Build eagerly builds an Image on Modal.
func (image *Image) Build(ctx context.Context, app *App, params *ImageBuildParams) (*Image, error) {
	// Image is already hyrdated
	if image.ImageID != "" {
		return image, nil
	}

	image.client.logger.DebugContext(ctx, "Building image", "app_id", app.AppID)

	for _, currentLayer := range image.layers {
		if err := validateDockerfileCommands(currentLayer.commands); err != nil {
			return nil, err
		}
	}

	currentImageID := image.baseImageID
	imageBuilderVersion, err := image.client.getImageBuilderVersion(ctx, app.Environment)
	if err != nil {
		return nil, err
	}

	image.client.logger.DebugContext(ctx, "Image build", "app_id", app.AppID, "image_builder_version", imageBuilderVersion, "environment_name", app.Environment)

	for i, currentLayer := range image.layers {
		if err := ctx.Err(); err != nil {
			return nil, err
		}

		mergedSecrets, err := mergeEnvIntoSecrets(ctx, image.client, &currentLayer.env, &currentLayer.secrets)
		if err != nil {
			return nil, err
		}

		if err := hydrateSecrets(ctx, image.client, mergedSecrets); err != nil {
			return nil, err
		}

		var secretIds []string
		for _, secret := range mergedSecrets {
			secretIds = append(secretIds, secret.SecretID)
		}

		var gpuConfig *pb.GPUConfig
		if currentLayer.gpu != "" {
			var err error
			gpuConfig, err = parseGPUConfig(currentLayer.gpu)
			if err != nil {
				return nil, err
			}
		}

		var dockerfileCommands []string
		var baseImages []*pb.BaseImage

		if i == 0 && currentImageID != "" {
			dockerfileCommands = append([]string{"FROM base"}, currentLayer.commands...)
			baseImages = []*pb.BaseImage{pb.BaseImage_builder{
				DockerTag: "base",
				ImageId:   currentImageID,
			}.Build()}
		} else if i == 0 {
			dockerfileCommands = append([]string{fmt.Sprintf("FROM %s", image.tag)}, currentLayer.commands...)
			baseImages = []*pb.BaseImage{}
		} else {
			dockerfileCommands = append([]string{"FROM base"}, currentLayer.commands...)
			baseImages = []*pb.BaseImage{pb.BaseImage_builder{
				DockerTag: "base",
				ImageId:   currentImageID,
			}.Build()}
		}

		var imageRegistryConfig *pb.ImageRegistryConfig
		if i == 0 && currentImageID == "" && image.buildRegistryConfig != nil {
			imageRegistryConfig, err = image.buildRegistryConfig(ctx, image.client)
			if err != nil {
				return nil, err
			}
		}

		resp, err := image.client.cpClient.ImageGetOrCreate(
			ctx,
			pb.ImageGetOrCreateRequest_builder{
				AppId: app.AppID,
				Image: pb.Image_builder{
					DockerfileCommands:  dockerfileCommands,
					ImageRegistryConfig: imageRegistryConfig,
					SecretIds:           secretIds,
					GpuConfig:           gpuConfig,
					ContextFiles:        []*pb.ImageContextFile{},
					BaseImages:          baseImages,
				}.Build(),
				BuilderVersion: imageBuilderVersion,
				ForceBuild:     currentLayer.forceBuild,
			}.Build(),
		)
		if err != nil {
			return nil, err
		}

		result := resp.GetResult()

		if result == nil || result.GetStatus() == pb.GenericResult_GENERIC_STATUS_UNSPECIFIED {
			// Not built or in the process of building - wait for build
			lastEntryID := ""
			for result == nil {
				if err := ctx.Err(); err != nil {
					return nil, err
				}

				var err error
				result, lastEntryID, err = image.waitForBuildIteration(ctx, resp.GetImageId(), lastEntryID)
				if err != nil {
					return nil, err
				}
			}
		}

		switch result.GetStatus() {
		case pb.GenericResult_GENERIC_STATUS_FAILURE:
			return nil, RemoteError{fmt.Sprintf("Image build for %s failed with the exception:\n%s", resp.GetImageId(), result.GetException())}
		case pb.GenericResult_GENERIC_STATUS_TERMINATED:
			return nil, RemoteError{fmt.Sprintf("Image build for %s terminated due to external shut-down. Please try again.", resp.GetImageId())}
		case pb.GenericResult_GENERIC_STATUS_TIMEOUT:
			return nil, RemoteError{fmt.Sprintf("Image build for %s timed out. Please try again with a larger timeout parameter.", resp.GetImageId())}
		case pb.GenericResult_GENERIC_STATUS_SUCCESS:
			// Success, do nothing
		default:
			return nil, RemoteError{fmt.Sprintf("Image build for %s failed with unknown status: %s", resp.GetImageId(), result.GetStatus())}
		}

		// The new image becomes the base for the next layer
		currentImageID = resp.GetImageId()
	}

	image.ImageID = currentImageID
	image.client.logger.DebugContext(ctx, "Image build completed", "image_id", currentImageID)
	return image, nil
}

// Publish publishes this built Image under a stable name and tag.
// The name may include a tag as name:tag; if no tag is included, :latest is used.
func (image *Image) Publish(ctx context.Context, name string, params *ImagePublishParams) error {
	if params == nil {
		params = &ImagePublishParams{}
	}
	namespacePrefix, tag, err := parseNamedImageRef(name)
	if err != nil {
		return err
	}
	if image.ImageID == "" {
		return InvalidError{Exception: "Cannot publish an image that has not been built yet. Call Build() first."}
	}

	var environmentName string
	if namespacePrefix != "" {
		if params.Environment != "" {
			return InvalidError{Exception: "Cannot specify 'Environment' when the image name contains a '/'."}
		}
		environmentName = ""
	} else {
		environmentName = firstNonEmpty(params.Environment, image.client.profile.Environment)
	}

	_, err = image.client.cpClient.ImagePublish(ctx, pb.ImagePublishRequest_builder{
		ImageId:         image.ImageID,
		EnvironmentName: environmentName,
		AllowPublic:     false,
		Tag:             tag,
	}.Build())
	return err
}

// ImageDeleteParams are options for deleting an Image.
type ImageDeleteParams struct{}

// Delete deletes an Image by ID.
//
// Deletion is irreversible and will prevent Functions/Sandboxes from using the Image.
//
// Note: When building an Image, each chained method call will create an
// intermediate Image layer, each with its own ID. Deleting an Image will not
// delete any of its intermediate layers, only the image identified by the
// provided ID.
func (s *imageServiceImpl) Delete(ctx context.Context, imageID string, params *ImageDeleteParams) error {
	_, err := s.client.cpClient.ImageDelete(ctx, pb.ImageDeleteRequest_builder{ImageId: imageID}.Build())
	if status, ok := status.FromError(err); ok && status.Code() == codes.NotFound {
		return NotFoundError{fmt.Sprintf("Image '%s' not found", imageID)}
	}
	return err
}
